File: dmoj/python_cpp/coci20c1p1.py
def dir_check(i, j, grid):
    d = 0
    while True:
        cur = grid[i][j]
        if cur == "." or cur == "o":
            return None

        d += 1
        if cur == ">":
            j += 1
        elif cur == "<":
            j -= 1
        elif cur == "^":
            i -= 1
        elif cur == "v":
            i += 1
        else:  # cur = 'x'
            return d


r, s = map(int, input().split())
grid = [input() for _ in range(r)]

for x0 in range(r):
    if "o" in grid[x0]:
        y0 = grid[x0].index("o")
        break

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
dirChar = ["N", "S", "W", "E"]


out = []
for k in range(4):
    nx = x0 + dx[k]
    ny = y0 + dy[k]
    d = dir_check(nx, ny, grid)
    if d is not None:
        out.append((d, dirChar[k]))

if not out:
    print(":(")
else:
    print(":)")
    print(min(out)[1])

File: dmoj/python_cpp/coci20c1p1.cpp
#include <algorithm>
#include <iostream>
#include <optional>
#include <string>
#include <utility>
#include <vector>

using namespace std;

optional<int> dir_check(int i, int j, const vector<string> &grid) {
  int d = 0;

  while (true) {
    char cur = grid[i][j];
    if (cur == '.')
      break;

    d++;
    if (cur == '>')
      j++;
    else if (cur == '<')
      j--;
    else if (cur == '^')
      i--;
    else if (cur == 'v')
      i++;
    else
      return d; // cur = 'x'
  }
  return nullopt;
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int r, s;
  cin >> r >> s;

  vector<string> grid(r);
  for (int i = 0; i < r; i++)
    cin >> grid[i];

  int x0 = -1, y0 = -1;
  for (int i = 0; i < r; i++) {
    auto pos = grid[i].find('o');
    if (pos != string::npos) {
      x0 = i;
      y0 = pos;
      break;
    }
  }

  constexpr int dx[4] = {-1, 1, 0, 0};
  constexpr int dy[4] = {0, 0, -1, 1};
  constexpr char dirChar[4] = {'N', 'S', 'W', 'E'};

  vector<pair<int, char>> out;

  for (int k = 0; k < 4; k++) {
    int nx = x0 + dx[k];
    int ny = y0 + dy[k];
    auto d = dir_check(nx, ny, grid);
    if (d.has_value()) {
      out.push_back({*d, dirChar[k]});
    }
  }

  if (out.empty()) {
    cout << ":(";
  } else {
    auto best = min_element(out.begin(), out.end());
    cout << ":)\n";
    cout << best->second;
  }

  return 0;
}
